Public/Src/Engine/Scheduler/Distribution/Worker.cs

```diff
@@ -421,27 +421,54 @@ internal bool TryAcquire(RunnablePip runnablePip, out WorkerResource? limitingRe
                 loadFactor = 1;
             }
 
-            if (AcquiredProcessSlots >= (EffectiveTotalProcessSlots * loadFactor))
+            var processRunnablePip = runnablePip as ProcessRunnablePip;
+            var acquiredSlots = m_acquiredProcessSlots;
+            var totalSlotsNeeded = acquiredSlots + processRunnablePip.Process.Weight;

```

Comment:
This might be nicer abstracted in a method `getNrOfSlotsForPip(runnablePip)`

Reply:
I am going to keep it here purely because this area is race-condition prone and I'd rather have all the possible interacting variables as transparent as possible. 

(There is actually still a race condition between reading EffectiveTotalProcessSlots here and the CompareExchange with m_acquiredProcessSlots, but it seemed relatively niche and I didn't want to add a lock to prevent it).

---
In reply to: [271090931](https://github.com/Microsoft/BuildXL/pull/67#discussion_r271090931) [](ancestors = 271090931)

Reply:
np.

---
In reply to: [271096310](https://github.com/Microsoft/BuildXL/pull/67#discussion_r271096310) [](ancestors = 271096310,271090931)